desktop/src/preferences/write.rs

```diff
@@ -1,146 +1,172 @@
+            }
+            values.output_device = name;
+        })
     }
 
     pub fn set_mute(&mut self, mute: bool) {
-        self.0.toml_document["mute"] = value(mute);
-        self.0.values.mute = mute;
+        self.0.edit(|values, toml_document| {
+            toml_document["mute"] = value(mute);
+            values.mute = mute;
+        })
     }
 
     pub fn set_volume(&mut self, volume: f32) {
-        self.0.toml_document["volume"] = value(volume as f64);
-        self.0.values.volume = volume;
+        self.0.edit(|values, toml_document| {
+            toml_document["volume"] = value(volume as f64);
+            values.volume = volume;
+        })
     }
 
     pub fn set_log_filename_pattern(&mut self, pattern: FilenamePattern) {
-        self.0.toml_document["log"]["filename_pattern"] = value(pattern.as_str());
-        self.0.values.log.filename_pattern = pattern;
+        self.0.edit(|values, toml_document| {
+            toml_document["log"]["filename_pattern"] = value(pattern.as_str());
+            values.log.filename_pattern = pattern;
+        })
     }
 
     pub fn set_storage_backend(&mut self, backend: StorageBackend) {
-        self.0.toml_document["storage"]["backend"] = value(backend.as_str());
-        self.0.values.storage.backend = backend;
+        self.0.edit(|values, toml_document| {
+            toml_document["storage"]["backend"] = value(backend.as_str());
+            values.storage.backend = backend;
+        })
     }
 }
 
-pub struct BookmarksWriter<'a>(&'a mut BookmarksAndDocument);
+pub struct BookmarksWriter<'a>(&'a mut DocumentHolder<Bookmarks>);
 
 impl<'a> BookmarksWriter<'a> {
-    pub(super) fn new(bookmarks: &'a mut BookmarksAndDocument) -> Self {
+    pub(super) fn new(bookmarks: &'a mut DocumentHolder<Bookmarks>) -> Self {
         Self(bookmarks)
     }
 
-    fn bookmark_table(&mut self, index: usize) -> &mut Table {
-        self.get_underlying_table()
-            .get_mut(index)
-            .expect("invalid bookmark index")
-    }
+    fn with_underlying_table(&mut self, fun: impl FnOnce(&mut Bookmarks, &mut ArrayOfTables)) {
+        fn find_table(toml_document: &mut DocumentMut) -> &mut ArrayOfTables {
+            if toml_document.contains_array_of_tables("bookmark") {
+                return toml_document["bookmark"]
+                    .as_array_of_tables_mut()
+                    .expect("type was just verified");
+            }
 
-    fn get_underlying_table(&mut self) -> &mut ArrayOfTables {
-        if self.0.toml_document.contains_array_of_tables("bookmark") {
-            return self.0.toml_document["bookmark"]
+            tracing::warn!("missing or invalid bookmark array, recreating..");
+            toml_document.insert("bookmark", array());
+            toml_document["bookmark"]
                 .as_array_of_tables_mut()
-                .expect("type was just verified");
+                .expect("type was just created")
         }
```

Comment:
side note:  not related to the PR but I thought about making a WriteExt and have this as a `get_or_create_...` method there.

Reply:
That'd probably look quite nice! But I'll save it for a followup, yeah? :D